src/cmd/tools/doctest/doctest.go

```diff
@@ -0,0 +1,193 @@
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package main
+
+import (
+	"bytes"
+	"flag"
+	"fmt"
+	"go/build"
+	"io/ioutil"
+	"log"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strings"
+)
+
+const (
+	operationRegEx  = `(?s)<operation>.*?</operation>|(?s)<operation_json>.*?</operation_json>`
+	validationRegEx = `(?s)<validation>.*?</validation>`
+
+	operationOpen      = `<operation>`
+	operationClose     = `</operation>`
+	operationJSONOpen  = `<operation_json>`
+	operationJSONClose = `</operation_json>`
+	validationOpen     = `<validation>`
+	validationClose    = `</validation>`
+
+	all = `<operation>|</operation>|<validation>|</validation>`
+)
+
+var (
+	file           string
+	destDir        string
+	markdown, bash bool
+
+	goPath = build.Default.GOPATH + "/src/"
+)
+
+func init() {
+	flag.StringVar(&file, "file", "", "File to be parsed")
+	flag.StringVar(&destDir, "dest", "", "Directory where the output will be saved")
+
+	flag.BoolVar(&markdown, "markdown", true, "generate a markdown file based on the source (default=true)")
+	flag.BoolVar(&bash, "bash", true, "generate a bash script based on the source (default=true)")
+}
+
+func main() {
+	flag.Parse()
+
+	// this is not ideal
+	if file == "" {
+		panic("'file' flag is required")
+	}
+
+	if destDir == "" {
+		destDir = filepath.Dir(file)
+	}
+
+	f, err := ioutil.ReadFile(file)
+	if err != nil {
+		log.Fatalf("unable to read source file: %v\n", err)
+	}
+
+	operationRegEx := regexp.MustCompile(operationRegEx)
+	validationRegEx := regexp.MustCompile(validationRegEx)
+
+	if bash {
+		if err := bashScript(f, destDir, file, operationRegEx, validationRegEx); err != nil {
+			log.Fatalf("unable to create bash script: %v", err)
+		}
+	}
+
+	if markdown {
+		if err := markdownFile(f, destDir, file, validationRegEx, operationRegEx); err != nil {
+			log.Fatalf("unable to create markdown file: %v", err)
+		}
+	}
+}
+
+func markdownFile(contents []byte, destDir, fileName string, validationRegEx *regexp.Regexp, opRegex *regexp.Regexp) error {
+	// create new markdown file based on name of source file
+	mdFileName := destDir + "/" + strings.Replace(filepath.Base(fileName), ".source", "", -1)
+	mdFile, err := os.Create(mdFileName)
```

Comment:
nit: you can use OpenFIle call to do the same thing as Create+Chmod. something like:

```
mdFile, err := os.OpenFile(name, os.O_RDWR|os.O_CREATE|os.O_TRUNC, os.FileMode(0644))
```

Reply:
I think this is fine for the markdown file